### ISSUE TITLE
threads.lua:183: [thread 4 callback] bad argument #2 to '?'

### ISSUE THREAD

**Issue by zmlmanly** [SeungjunNah/DeepDeblur_release#9]
Hi, I am training my model by your shared code, but I have a problem. When I use "th main.lua -nEpochs 450 -save scale3", the result is :
==> online epoch #1 [mini-batchsize = 4]
/root/torch/install/bin/luajit : threads.lua:183: [thread 4 callback] bad argument #2 to '?' (cannot   sample n_sample > prob_dist:size(1) samples without replacement at /root/torch/pkg/torch/lib/TH/generic/THTensorRandom.c:204)
How to solver the problem?

**Comment by SeungjunNah** [comment]
Hi,
I cannot reproduce the same error message.
Did it happen after training for a while? or immediately after you launched process?
From your error message, it seems like error happened during multinomial random sampling (line 139).
It indicates that the length of variable "train_list" is less than "batch_size" which is 4 by default.
Did you place the dataset correctly?
It should look like: 
"DeepDeblur_release/dataset/GOPRO_Large/train/GOPR0372_07_00/blur_gamma/000047.png"

**Comment by zmlmanly** [comment]
Thank you for your help. I have check the length of "train_list", and it is correct. The dataset is also same to you. It seems like the Threads error.

**Comment by SeungjunNah** [comment]
You can try debugging by setting the number of threads as 0. (single threaded data loading)
"th main.lua -nDonkeys 0".
If it is safe when opt.nDonkeys == 0, then it is most likely Threads problem.

**Comment by zmlmanly** [comment]
OK，Thank you very much. I will try it again.

**Comment by zmlmanly** [comment]
@SeungjunNah , That problem has solved, but I have another problem. 
train.lua:185  model_container:backward(blur, criterion_container.gradInput)
It shows the error: ConcatTable.lua:37: currentGradInput is not table!
How to solve it? Thank you very much.

**Comment by SeungjunNah** [comment]
Could you explain more about In which case the error occurred?
And would you show the full error message?

**Comment by zmlmanly** [comment]
ok, I am sorry about that. Print information:
.....
epoch begins:1
==>training!
==>doing epoch on training data:
==>online epoch # 1 [mini-batchSize = 4]
==>here is the loss function:
nn.ParallelCriterion
/root/torch/install/bin/luajit:/root/torch/install/share/lua/5.1/nn/Container.lua:67:
In 2 module of nn.Sequential:
/root/torch/install/share/lua/5.1/nn/ConcatTable.lua:37:currentGradInput is not table!
stack traceback:
               [c]:in fuction 'error'
               /root/torch/install/share/lua/5.1/nn/ConcatTable.lua:37: in function </root/torch/install/share/lua/5.1/nn/ConcatTable.lua:30>
              [c]:in fuction 'xpcall'
               /root/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
               /root/torch/install/share/lua/5.1/nn/Sequential.lua:84: in function 'backward'
              train.lua:185 :in function 'opfunc'
               /root/torch/install/share/lua/5.1/optim/adam.lua:37: in function 'G'
              train.lua:214 :in function 'trainBatch'
              ......

**Comment by SeungjunNah** [comment]
Your log says there's a problem updating discriminator model, which is also called as  model_container:get(2)
The gradInput should look like:
```lua
th> model_container:get(2).gradInput
{
  1 : CudaTensor - size: 4x3x256x256
  2 : CudaTensor - size: 4x3x128x128
  3 : CudaTensor - size: 4x3x64x64
}
```
It is then fed into nn.SelectTable(), becoming CudaTensor - size: 4x3x256x256
I have no idea why your log says there's a problem.
Did you modify any source code such as model.lua or loss.lua?

And what is your model_container.gradInput?
You can print them by inserting 
```lua
require 'trepl.lua'()
```
into some line and try printing in interactive mode.
It should look like
```lua
th> model_container.gradInput
{
  1 : CudaTensor - size: 4x3x256x256
  2 : CudaTensor - size: 4x3x128x128
  3 : CudaTensor - size: 4x3x64x64
}

```

**Comment by zmlmanly** [comment]
@SeungjunNah  I am sorry to reply now. I do not modify any source code such as model.lua or loss.lua. I insert require 'trepl'() in feval.G = function(x), and I find "loss = criterion_container(output, gt)" error. I print the output and gt, is 
th> output
{
1:
{
  1 : CudaTensor - size: 4x3x256x256
  2 : CudaTensor - size: 4x3x128x128
  3 : CudaTensor - size: 4x3x64x64
}
2: CudaTensor -size: 4x1x1x1
}
th> gt
{
1:
{
  1 : CudaTensor - size: 4x3x256x256
  2 : CudaTensor - size: 4x3x128x128
  3 : CudaTensor - size: 4x3x64x64
}
2: CudaTensor -size: 4
}
Is the result same with your? Why print the error?

 And when I run criterion.D(output_label, false_label), it will print error:
/root/torch/install/share/lua/5.1/nn/BCECriterion.lua:33: attempt to call field 'BCECriterion_updateOutput' (a nil value)
But I print the output_label and false_label, 
th> output_label
(1,1,.,.) = 
0.5019
(2,1,.,.) = 
0.5019
(3,1,.,.) = 
0.5019
(4,1,.,.) = 
0.5019
[torch.CudaTensor of size 4x1x1x1]

th>  false_label
0
0
0
0
[torch.CudaTensor of size 4]
Absolutely, there are some values in output_label and false_label, why print the error "'BCECriterion_updateOutput' (a nil value)"?

The version of lua that I use is 5.1。

**Comment by SeungjunNah** [comment]
I have the same output, gt, output_label, and false_label variables as you.
It seems that you have the desired input and target formats for the loss function.
However, it looks strange that you got error log saying 'BCECriterion_updateOutput' is not found.
nn.BCECriterion is a built-in loss, it shouldn't cause problems as long as the installation is correct.
Could you try updating or re-installing torch?
```bash
cd torch
./update.sh
```
If everything's fine with torch, I have no idea.

**Comment by zmlmanly** [comment]
ok, Thank you for your help. I'll try it again.

**Comment by zmlmanly** [comment]
@SeungjunNah  I doubt the torch which is installed is worng. I install torch7 through the methods:
(1)git clone https://github.com/torch/distro.git ~/torch --recursive
(2)cd ~/torch
(3)bash install-deps
(4)./install.sh
What is your methods to install torch7?
And my cuda-version is V8.0.44, what is your cuda-version?

**Comment by SeungjunNah** [comment]
I have used every version of cuda since 7.5 to current 9.0 to run this code and had no problem using it.
Could you try this then?
If this script gives you the error, then torch is the problem. If not, we should look into the source code again.
```lua
require 'torch'
require 'nn'
require 'cunn'
require 'cudnn'

weights = torch.Tensor(4):fill(1/torch.log(2))
criterion = nn.BCECriterion(weights):cuda()
output_label = torch.CudaTensor(4)
false_label = torch.CudaTensor(4):zero()

loss = criterion(output_label, false_label)
```

By the way, while trying above script, when I omitted
```lua
require 'cunn'
```
in the above script, it gave me similar (but not same) error as you,  "...seungjun/torch/install/share/lua/5.1/nn/BCECriterion.lua:33: attempt to index field 'THNN' (a nil value)" when executing
```lua
loss = criterion(output_label, false_label)
```

Could it be the case that inserting 
```lua
require 'cunn'
```
to appropriate places (i.e. somewhere in train.lua) helps you?

**Comment by Harry-675** [comment]
@zmlmanly Have you success running the process?

**Comment by ttyyteresa** [comment]
@zmlmanly I have the same problem.Have you solve it

**Comment by harrytruong0804** [comment]
I have the same problem too. Have you all solve it?

**Comment by SeungjunNah** [comment]
@ttyyteresa @noitq 
So far I couldn't reproduce the error. I'm sorry.

**Comment by SeungjunNah** [comment]
PyTorch version now available here: https://github.com/SeungjunNah/DeepDeblur-PyTorch
